python/pyarrow/tests/test_cuda_numba_interop.py

```diff
@@ -0,0 +1,140 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+import pytest
+import pyarrow as pa
+import numpy as np
+
+dtypes = ['uint8', 'int16', 'float32']
+cuda = pytest.importorskip("pyarrow.cuda")
+nb_cuda = pytest.importorskip("numba.cuda")
+
+from numba.cuda.cudadrv.devicearray import DeviceNDArray  # noqa: E402
+
+
+ctx = None
+nb_ctx = None
+
+
+def setup_module(module):
+    # It turns out that for numba-pyarrow.cuda interop one should
+    # create CUDA context using numba context manager.
+
+    if 0:
+        # At some point tests will fail
```

Comment:
These are good questions and I don't have answers to these.

Here are the symptoms:
Making pyarrow.cuda context manager effective (by changing `if 0:` to `if 1:`) will lead to test failures. The test failures actually depend on the order of running the tests: there exists ordering when all tests succeed. However, I would not reorder the tests for hiding the issue here. 

My perception is that numba.cuda context manager is not designed to handle context objects created by foreign libraries. In order to make the numba-pyarrow interoperability to work with pyarrow.cuda context manager, one might need to modify numba but that would be out of the scope of this PR.

Reply:
Do you have a precise idea of what doesn't work? It would be nice to note that in a comment.
Also, please try to report the issue to the Numba developers on their bug tracker. We would like things to improve here :-)

Reply:
Sure, I understand the goal. However, currently I can only observe the first failing test showing a message:
```
Call to cuLaunchKernel results in CUDA_ERROR_INVALID_HANDLE
```
and in the following failing tests numba.cuda context manager reports a mismatch of context handle values when popping the context.
I have tried to track down the issue within numba but without success so far...

Reply:
Well, adding that in a comment sounds already more precise than """At some point tests will fail""" :-)

Also, please report a bug to the Numba devs. They could help you investigate, or even investigate by themselves.

Reply:
A perhaps related issue: https://github.com/numba/numba/issues/3247

Reply:
I don't know, it doesn't seem obvious to me. Best is to report the issue separately, IMO.

Reply:
Created https://github.com/numba/numba/issues/3410